internal/service/applications_data_source.go

```diff
@@ -0,0 +1,202 @@
+package service
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+
+	"terraform-provider-coolify/internal/api"
+	"terraform-provider-coolify/internal/filter"
+	"terraform-provider-coolify/internal/flatten"
+	"terraform-provider-coolify/internal/provider/generated/datasource_applications"
+	"terraform-provider-coolify/internal/provider/util"
+)
+
+var _ datasource.DataSource = &applicationsDataSource{}
+var _ datasource.DataSourceWithConfigure = &applicationsDataSource{}
+
+func NewApplicationsDataSource() datasource.DataSource {
+	return &applicationsDataSource{}
+}
+
+type applicationsDataSource struct {
+	client *api.ClientWithResponses
+}
+
+type applicationsDataSourceWithFilterModel struct {
+	datasource_applications.ApplicationsModel
+	Filter []filter.BlockModel `tfsdk:"filter"`
+}
+
+var applicationsFilterNames = []string{"id", "uuid", "name", "description", "fqdn"}
+
+func (d *applicationsDataSource) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_applications"
+}
+
+func (d *applicationsDataSource) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = datasource_applications.ApplicationsDataSourceSchema(ctx)
+	resp.Schema.Description = "Get a list of Coolify applications."
+
+	// todo: Mark sensitive attributes
```

Comment:
What approach would be better here for this? Since the `makeDataSourceAttributeSensitive` is currently primarily designed for only top level attributes, like the application data source, and not plural data sources like `projects` or `servers` where the actual attributes will be nested in a sub-attribute.

Would it be better to make a 2nd function that duplicates the logic of `makeDataSourceAttributeSensitive` but is designed to handle nested attributes or just update the existing `makeDataSourceAttributeSensitive` function to take an optional 3rd parameter to define a nested attribute to apply to?

Reply:
`makeDataSourceAttributeSensitive`'s first parameter is any datasource attribute map, there is no restriction on being top-level or not.
Go is strongly typed, so you'll need to do a [type assertion](https://go.dev/tour/methods/15) (narrowing in TypeScript) on the `applications` attribute,  in order to confidently pass the nested object's attributes to the `makeDataSourceAttributeSensitive` function.
Once that is done, run `make generate` again to update the docs and the attributes should be marked as sensitive.

<details>
<summary>(skip to the solution)</summary>

```diff
diff --git a/internal/service/applications_data_source.go b/internal/service/applications_data_source.go
index 7f50a35..07611df 100644
--- a/internal/service/applications_data_source.go
+++ b/internal/service/applications_data_source.go
@@ -44,7 +44,14 @@ func (d *applicationsDataSource) Schema(ctx context.Context, req datasource.Sche
 	resp.Schema = datasource_applications.ApplicationsDataSourceSchema(ctx)
 	resp.Schema.Description = "Get a list of Coolify applications."
 
-	// todo: Mark sensitive attributes
+	// Mark sensitive attributes
+	sensitiveAttrs := []string{"manual_webhook_secret_bitbucket", "manual_webhook_secret_gitea", "manual_webhook_secret_github", "manual_webhook_secret_gitlab"}
+	if attrSet, ok := resp.Schema.Attributes["applications"].(schema.SetNestedAttribute); ok {
+		for _, attr := range sensitiveAttrs {
+			makeDataSourceAttributeSensitive(attrSet.NestedObject.Attributes, attr)
+		}
+	}
+
 	resp.Schema.Blocks = map[string]schema.Block{
 		"filter": filter.CreateDatasourceFilter(applicationsFilterNames),
 	}

```
</details>